src/filesystem.rs

```diff
@@ -47,175 +47,319 @@ pub struct FilesystemReader<R: SquashFsReader> {
+        FilesystemReaderFile::new(self, basic_file)
     }
 
     /// Read and return all the bytes from the file
     pub fn read_file(&self, basic_file: &BasicFile) -> Result<Vec<u8>, SquashfsError> {
-        let mut reader = FilesystemFileReader::new(self, basic_file);
+        let file = FilesystemReaderFile::new(self, basic_file);
+        let mut reader = file.raw_data_reader().decompress().into_reader();
         let mut bytes = Vec::with_capacity(basic_file.file_size as usize);
         reader.read_to_end(&mut bytes)?;
         Ok(bytes)
     }
-
-    /// Read from either Data blocks or Fragments blocks
-    fn read_data(&self, size: usize) -> Result<Vec<u8>, SquashfsError> {
-        let uncompressed = size & (DATA_STORED_UNCOMPRESSED as usize) != 0;
-        let size = size & !(DATA_STORED_UNCOMPRESSED as usize);
-        let mut buf = vec![0u8; size];
-        self.reader.borrow_mut().read_exact(&mut buf)?;
-
-        let bytes = if uncompressed {
-            buf
-        } else {
-            compressor::decompress(buf, self.compressor)?
-        };
-        Ok(bytes)
-    }
 }
 
-struct FilesystemFileReader<'a, R: SquashFsReader>(Option<InnerFilesystemFileReader<'a, R>>);
-impl<'a, R: SquashFsReader> FilesystemFileReader<'a, R> {
-    pub fn new(filesystem: &'a FilesystemReader<R>, file: &'a BasicFile) -> Self {
-        Self(Some(InnerFilesystemFileReader {
-            filesystem,
-            file,
-            last_read: vec![],
-            current_block: Some(0),
-            bytes_available: file.file_size as usize,
-            pos: file.blocks_start.into(),
-        }))
+#[derive(Clone, Copy)]
+pub struct FilesystemReaderFile<'a, R: ReadSeek> {
+    system: &'a FilesystemReader<R>,
+    basic: &'a BasicFile,
+}
+impl<'a, R: ReadSeek> FilesystemReaderFile<'a, R> {
+    pub fn new(system: &'a FilesystemReader<R>, basic: &'a BasicFile) -> Self {
+        Self { system, basic }
+    }
+    pub fn reader(&self) -> SquashfsReadFile<'a, R> {
+        self.raw_data_reader().decompress().into_reader()
+    }
+    pub fn raw_data_reader(&self) -> SquashfsRawData<'a, R> {
+        SquashfsRawData::new(Self {
+            system: self.system,
+            basic: self.basic,
+        })
     }
 }
-impl<'a, R: SquashFsReader> Read for FilesystemFileReader<'a, R> {
-    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
-        let inner = if let Some(inner) = &mut self.0 {
-            inner
-        } else {
-            return Ok(0);
+enum BlockFragment<'a> {
+    Block(&'a DataSize),
+    Fragment(&'a Fragment),
+}
+pub enum RawDataBlock<'a, 'b> {
+    Block {
+        data: &'b [u8],
+        block: &'a DataSize,
+    },
+    Fragment {
+        data: &'b [u8],
+        fragment: &'a Fragment,
+    },
+    CachedFragment {
+        data: &'b [u8],
+        fragment: Fragment,
+    },
+}
+//TODO: DISCUSTING, please have mercy on my soul
```

Comment:
lmao

Reply:
I need to find a more elegant solution for that, this is horrible. I'll try to fix this in the next few days.